File: elixir/problem_1.exs
defmodule Problem1 do
  @moduledoc """
    This module is responsible for solving Problem 1 of Project Euler.
    It should find the sum of all the natural numbers divisible by 3 or 5 that are less than x (default value of 1000).
  """
  
  @spec problem_1(non_neg_integer()) :: non_neg_integer()
  def problem_1(x \\ 1000), do: Enum.reduce(0..x - 1, &add/2)

  defp add(num, sum) when rem(num, 3) == 0 or rem(num, 5) == 0, do: num + sum
  defp add(_, sum), do: sum
end

ExUnit.start()

defmodule Day1Test do
  use ExUnit.Case

  import Problem1

  test "sum all multiples of 3 or 5 below 1000" do
    assert problem_1() == 233168
  end
end

File: javascript/problem1.js
/**
 * @description Find the sum of all the multiples of 3 or 5 below upperBound (default value of 1000).
 * @param {integer} upperBound
 * @returns {integer}
 */

function problem1(upperBound = 1000) {
  let sum = 0;
  for (let i = 0; i < upperBound; i +=1) {
    if (i % 3 == 0 || i % 5 == 0) sum += i;
  }
  return sum;
}

console.log(`the sum of all the multiples of 3 or 5 below 1000 is ${problem1()}.`)
